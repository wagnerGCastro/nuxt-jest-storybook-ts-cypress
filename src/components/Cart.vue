<template>
  <div class="cart">
    <hr class="space-hr mb-6" />

    <v-row v-for="(p, idx) in 5" :key="idx" no-gutters class="mb-3">
      <v-col cols="4" class="mb-4">
        <div class="box-image">
          <v-card outlined>
            <v-img
              lazy-src="https://picsum.photos/id/11/10/6"
              src="https://picsum.photos/id/11/500/300"
            ></v-img>
          </v-card>
          <button data-testid="remove-button" class="remove-product">X</button>
        </div>
      </v-col>

      <v-col cols="8">
        <div class="d-flex items-center">
          <div class="ml-3">
            <h3 class="title-product">DISTINCTIO EXCEPTURI3434 DOLORUM</h3>
            <div class="d-flex items-center justify-space-between mt-1">
              <div class="d-flex items-center">
                <svg
                  fill="none"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  stroke="currentColor"
                  class="decrase"
                >
                  <path d="M15 12H9m12 0a9 9 0 11-18 0 9 9 0 0118 0z"></path>
                </svg>

                <span data-testid="quantity" class="quantity pl-2 pr-2">10</span>

                <svg
                  fill="none"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  stroke="currentColor"
                  class="plus"
                >
                  <path d="M12 9v3m0 0v3m0-3h3m-3 0H9m12 0a9 9 0 11-18 0 9 9 0 0118 0z"></path>
                </svg>
              </div>

              <span class="price">R$ 80,00</span>
            </div>
          </div>
        </div>
      </v-col>
    </v-row>

    <hr class="space-hr" />

    <div data-testid="checkout-form" @submit.prevent="checkout">
      <div class="d-flex justify-space-between">
        <p cols="4">Subtotal</p>
        <p cols="8" class="subtotal-price"><span class="text-xs">R$</span>899,00</p>
      </div>

      <div class="d-flex justify-center">
        <v-btn data-testid="checkout-button" type="submit" class="v-btn__button">
          <span>Checkout</span>
          <v-icon light> mdi-arrow-right-thin </v-icon>
        </v-btn>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'Cart',
  data() {
    return {
      //
    };
  },
  watch: {
    //
  },
  methods: {},
};
</script>

<style scoped>
.cart {
  padding: 0 16px;
}

.cart .decrase,
.cart .plus {
  width: 20px;
  color: #939393;
}

.cart .box-image {
  position: relative;
}

.remove-product {
  position: absolute;
  top: -7px;
  border: 1px solid #988;
  border-radius: 37px;
  height: 16px;
  width: 16px;
  font-size: 10px;
  line-height: 5px;
  left: -8px;
  color: #939393;
}

.remove-product:hover {
  border: 1px solid #494242;
  color: #494242;
}

.v-btn__button {
  width: 100%;
}

.title-product {
  font-size: 12px;
  font-weight: 400;
}

.price {
  font-size: 14px;
  min-width: 60px;
  text-align: right;
  font-weight: 500;
}

.quantity {
  font-size: 14px;
}

.space-hr {
  margin: auto 0 20px;
  opacity: 0.3;
}

.subtotal-price {
  color: blue;
}

.text-xs {
  font-size: 12px;
  margin-right: 3px;
}
</style>
