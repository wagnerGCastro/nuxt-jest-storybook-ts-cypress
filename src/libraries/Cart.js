import find from 'lodash/find';
import remove from 'lodash/remove';
import Dinero from 'dinero.js';

const Money = Dinero;
Money.defaultCurrency = 'BRL';
Money.defaultPrecision = 2;

export default class Cart {
  items = [];

  add(item) {
    const itemToFind = { product: item.product };

    if (itemToFind) {
      if (find(this.items, itemToFind)) {
        remove(this.items, itemToFind);
      }
      this.items.push(item);
    }
  }

  remove(product) {
    remove(this.items, { product });
  }

  getTotal() {
    return this.items.reduce((acc, { condition, product, quantity }) => {
      const isEven = quantity % 2 === 0;
      const amount = Money({ amount: quantity * product.price });
      let discount = Money({ amount: 0 });

      // // Calculate Percentage Discount
      // if (item.condition?.percentage && item.quantity > item.condition.minimum) {
      //   discount = amount.percentage(item.condition.percentage)
      // }
      // // Calculate Quantity Discount
      // else if (item.condition?.quantity && item.quantity > item.condition.quantity) {
      //   discount = amount.percentage(isEven ? 50 : 40)
      // }

      if (condition) {
        const list = Array.isArray(condition) ? condition : [condition];

        const [heigherDiscount] = list
          .map((c) => {
            let result = [discount];

            // Calculate Quantity Discount
            if (c.percentage && quantity > c.minimum) {
              result = amount.percentage(c.percentage).getAmount();
            }
            // Calculate Quantity Discount
            else if (c.quantity && quantity > c.quantity) {
              result = amount.percentage(isEven ? 50 : 40).getAmount();
            }

            return result;
          })
          .sort((a, b) => b - a);

        discount = Money({ amount: heigherDiscount });
      }

      return acc.add(amount).subtract(discount);
    }, Money({ amount: 0 }));
  }

  summary() {
    const total = this.getTotal();
    const formatted = total.toFormat('$0,0.00');
    const items = this.items;

    return {
      total: total.getAmount(),
      formatted,
      items,
    };
  }

  checkout() {
    const { total, items } = this.summary();
    this.items = [];

    return {
      total,
      items,
    };
  }
}
