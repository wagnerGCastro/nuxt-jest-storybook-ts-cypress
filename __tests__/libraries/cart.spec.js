import Cart from '~/libraries/Cart'

describe('Cart', () => {
  let cart

  const product = {
    title: 'Adidas running shoes - men',
    price: 35388,
  }

  const product2 = {
    title: 'Adidas running shoes - women',
    price: 41872,
  }

  beforeEach(() => {
    cart = new Cart()
  })

  describe('getTotal()', () => {
    it('should return 0 when getTotal() is executed in a newly created instance', () => {
      expect(cart.getTotal()).toEqual(0)
    })

    it('should multiply quantity and price and receive the toatal amount', () => {
      const item = {
        product: {
          title: 'Adidas runnig shoes - men',
          price: 35388, // 353.88 | R$ 383,8
        },
        quantity: 2, // 2*35388=70776
      }

      cart.add(item)
      expect(cart.getTotal()).toEqual(70776)
    })

    it('should ensure no more than on product exists at a time', () => {
      cart.add({
        product,
        quantity: 2,
      })

      cart.add({
        product,
        quantity: 1,
      })

      expect(cart.getTotal()).toEqual(35388)
    })

    it('should update total when a product gets included and then removed', () => {
      cart.add({
        product,
        quantity: 2,
      })

      cart.add({
        product: product2,
        quantity: 1,
      })

      cart.remove(product)

      expect(cart.getTotal()).toEqual(41872)
    })
  })

  describe('checkout()', () => {
    it('should return an object with the total an the list of items', () => {
      cart.add({
        product,
        quantity: 2,
      })

      cart.add({
        product: product2,
        quantity: 4,
      })

      expect(cart.checkout()).toMatchSnapshot()
    })

    it('should return an object with the total an the list of items when summary() is called', () => {
      cart.add({
        product,
        quantity: 1,
      })

      cart.add({
        product: product2,
        quantity: 2,
      })

      expect(cart.summary()).toMatchSnapshot()
      expect(cart.getTotal()).toBeGreaterThan(0)
    })

    it('should reset the cart when checkout() is called', () => {
      cart.add({
        product: product2,
        quantity: 3,
      })

      cart.checkout()

      expect(cart.getTotal()).toEqual(0)
    })
  })
})
