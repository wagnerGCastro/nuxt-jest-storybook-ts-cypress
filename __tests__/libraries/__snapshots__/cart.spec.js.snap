// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout() should return an object with the total an the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas running shoes - men",
      },
      "quantity": 2,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Adidas running shoes - women",
      },
      "quantity": 4,
    },
  ],
  "total": 238264,
}
`;

exports[`Cart checkout() should return an object with the total an the list of items when summary() is called 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 35388,
        "title": "Adidas running shoes - men",
      },
      "quantity": 1,
    },
    Object {
      "product": Object {
        "price": 41872,
        "title": "Adidas running shoes - women",
      },
      "quantity": 2,
    },
  ],
  "total": 119132,
}
`;
